libsmtutil/Z3Interface.cpp

```diff
@@ -243,6 +244,91 @@ z3::expr Z3Interface::toZ3Expr(Expression const& _expr)
 	smtAssert(false, "");
 }
 
+Expression Z3Interface::fromZ3Expr(z3::expr const& _expr)
+{
+	auto sort = fromZ3Sort(_expr.get_sort());
+	if (_expr.is_const() || _expr.is_var())
+		return Expression(_expr.to_string(), {}, sort);
+
+	smtAssert(_expr.is_app(), "");
+	vector<Expression> arguments;
+	for (unsigned i = 0; i < _expr.num_args(); ++i)
+		arguments.push_back(fromZ3Expr(_expr.arg(i)));
+
+	auto kind = _expr.decl().decl_kind();
+	if (_expr.is_ite())
+		return Expression::ite(arguments[0], arguments[1], arguments[2]);
+	else if (_expr.is_not())
+		return !arguments[0];
+	else if (_expr.is_and())
+		return arguments[0] && arguments[1];
+	else if (_expr.is_or())
+		return arguments[0] || arguments[1];
+	else if (_expr.is_implies())
+		return Expression::implies(arguments[0], arguments[1]);
+	else if (_expr.is_eq())
+		return arguments[0] == arguments[1];
+	else if (kind == Z3_OP_ULT || kind == Z3_OP_SLT)
+		return arguments[0] < arguments[1];
+	else if (kind == Z3_OP_ULEQ || kind == Z3_OP_SLEQ)
+		return arguments[0] <= arguments[1];
+	else if (kind == Z3_OP_GT || kind == Z3_OP_SGT)
+		return arguments[0] > arguments[1];
+	else if (kind == Z3_OP_UGEQ || kind == Z3_OP_SGEQ)
+		return arguments[0] >= arguments[1];
+	else if (kind == Z3_OP_ADD)
+		return arguments[0] + arguments[1];
+	else if (kind == Z3_OP_SUB)
+		return arguments[0] - arguments[1];
+	else if (kind == Z3_OP_MUL)
+		return arguments[0] * arguments[1];
+	else if (kind == Z3_OP_DIV)
+		return arguments[0] / arguments[1];
+	else if (kind == Z3_OP_MOD)
+		return arguments[0] % arguments[1];
+	else if (kind == Z3_OP_XOR)
+		return arguments[0] ^ arguments[1];
+	else if (kind == Z3_OP_BSHL)
+		return arguments[0] << arguments[1];
+	else if (kind == Z3_OP_BLSHR)
+		return arguments[0] >> arguments[1];
+	else if (kind == Z3_OP_BASHR)
+		return Expression::ashr(arguments[0], arguments[1]);
+	else if (kind == Z3_OP_INT2BV)
+	{
+		smtAssert(false, "");
+		size_t size = std::stoul(arguments[1].name);
+		return Expression::int2bv(arguments[0], size);
```

Comment:
You left it here just in case? Right now it is unreachable code :)

Also maybe doing like this:
```suggestion
		return Expression::int2bv(arguments[0], stoul(arguments[1].name));
```

Reply:
Yea, I'll remove it to be more explicit. We don't know for sure what comes from z3 in such cases.